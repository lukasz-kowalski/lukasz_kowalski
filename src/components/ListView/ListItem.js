import React from 'react';
import { Link } from 'react-router-dom';

class ListItem extends React.Component {
  handleDelete = () => {
    this.props.deleteCharacter(this.props.character.id);
  }

  render() {
    const { character } = this.props;
    return (
      <tr key={character.name}>
        <th scope="row">{character.id}</th>
        <td>{character.name}</td>
        <td>{character.species}</td>
        <td>{character.gender}</td>
        <td>{character.homeworld}</td>
        <td>
          <div className="btn-group btn-group-sm" role="group" aria-label="Actions">
            <button type="button" className="btn btn-secondary">
              <Link 
                to={{
                  pathname: `/edit-character/${character.id}`,
                  state: character
                }}
              >
                <i className="fa fa-pencil" aria-hidden="true"/> Edit
              </Link>
            </button>
            <button type="button" className="btn btn-danger" onClick={this.handleDelete}>
              <i className="fa fa-trash-o" aria-hidden="true"/> Remove
            </button>
          </div>
        </td>
      </tr>
    );
  }
}

export default ListItem;
