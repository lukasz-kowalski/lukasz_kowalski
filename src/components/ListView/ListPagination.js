import React from 'react';
import { Link } from 'react-router-dom';

const ListPagination = ({ isFirstPage, isLastPage, numCurrentPage ,paginationData }) => (
   <nav aria-label="Data grid navigation">
    <ul className="pagination justify-content-end">
      <li className={`page-item ${isFirstPage ? 'disabled' : ''}`}>
        <Link 
          to={{ pathname: `/characters/${numCurrentPage - 1}`, search: `_page=${numCurrentPage - 1}`}}
          className="page-link" 
          tabIndex={isFirstPage ? '-1' : '0'}
        >
          Previous
        </Link>
      </li>

      {paginationData.map(link => link)}

      <li className={`page-item ${isLastPage ? 'disabled' : ''}`}>
        <Link 
          to={{ pathname: `/characters/${numCurrentPage + 1}`, search: `_page=${numCurrentPage + 1}`}}
          className="page-link" 
          tabIndex={isLastPage ? '-1' : '0'}
        >
          Next
        </Link>
      </li>
    </ul>
  </nav>
);


export default ListPagination;
