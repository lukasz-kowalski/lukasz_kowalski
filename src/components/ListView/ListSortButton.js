import React from 'react';

class ListSortButton extends React.Component {
  handleClick = () => {
    this.props.handleSortChange(this.props.name.toLowerCase());
  }

  render() {
    const { name } = this.props;
    return (
      <th scope="col">
      {name !== 'Actions' ?
        <button 
          className="th-button" 
          onClick={this.handleClick}
        >
          {name}
        </button> :
        name
      }
      </th>
    );
  }
}

export default ListSortButton;
