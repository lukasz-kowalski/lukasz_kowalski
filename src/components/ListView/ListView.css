.active {
  z-index: 1;
  color: #fff;
  background-color: #007bff;
  border-color: #007bff;
}

input[type="radio"], .radio-label {
  display: block;
}

th[scope="col"] {
  position: relative;
  height: 3rem;
  min-width: 4rem;
}

.th-button {
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  background: transparent;
  border: none;
  text-align: left;
  font-weight: bold;
  outline-offset: -3px;
  cursor: pointer;
}

.th-button:hover {
  background: #ddd;
}
