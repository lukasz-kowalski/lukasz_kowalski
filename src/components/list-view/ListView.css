.active {
  z-index: 1;
  color: #fff;
  background-color: #007bff;
  border-color: #007bff;
}
