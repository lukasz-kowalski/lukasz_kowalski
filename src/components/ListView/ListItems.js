import React from 'react';
import ListItem from './ListItem';
import ListSortButton from './ListSortButton';

const buttonNames = [
  'Id', 'Name', 'Species', 'Gender', 'Homeworld', 'Actions'
]

const ListItems = ({ characters, deleteCharacter, handleSortChange }) => (
  <table className="table table-bordered table-hover">
    <thead className="thead-light">
    <tr>
      {buttonNames.map(item => 
        <ListSortButton
          key={item}
          handleSortChange={handleSortChange}
          name={item}
        />
      )}
    </tr>
    </thead>
    <tbody>
    {characters.map(character =>
      <ListItem
        key={character.name + character.id}
        character={character} 
        deleteCharacter={deleteCharacter}
      />
    )}
    </tbody>
  </table>
);

export default ListItems;
