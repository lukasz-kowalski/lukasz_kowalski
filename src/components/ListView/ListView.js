import React, { Component } from 'react';
import { NavLink, Link } from 'react-router-dom';
import debounce from '../../utils/debounce';
import ListItems from './ListItems';
import ListPagination from './ListPagination';
import SearchInput from '../ui/SearchInput';

import './ListView.css';

class ListView extends Component {
  constructor(props) {
    super(props);
    this.fetchSearchData = debounce(this.fetchSearchData, 200);
  }

  state = {
    characters: [],
    currentPage: '',
    firstPage: '',
    lastPage: '',
    searchParam: '',
    sortParam: ''
  }

  componentDidMount() {
    this.fetchData();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.match.params.page !== this.props.match.params.page) {
      this.fetchData();
    }
  }

  getInitialQuery = () => {
    const { params } = this.props.match;
    const currentPage = params.page ? params.page : '1';
    const initialQuery = `_page=${currentPage}`;
    this.setState({
      currentPage
    })
    return initialQuery;
  }

  setSearchToQuery = () => {
    const initialQuery = this.getInitialQuery();
    const { searchParam } = this.state;
    if (searchParam) {
      return '_page=1' + searchParam;
    }
    return initialQuery;
  }

  setSortToQuery = () => {
    const queryWithSearch = this.setSearchToQuery();
    const { sortParam } = this.state;
    if (sortParam) {
      return queryWithSearch + sortParam;
    }
    return queryWithSearch;
  }

  setURL = retry => {
    const { params } = this.props.match;
    const { history } = this.props;
    const { pathname } = this.props.location;
    const currentPage = params.page ? params.page : '1';
    let newPathname = pathname;
    if (retry) {
      const pathArr = newPathname.split('/');
      pathArr[2] -= 1;
      newPathname = pathArr.join('/');
    }
    const redirectedPathname = newPathname === '/' ? `/characters/${currentPage}` : newPathname;
    const initialURL = `${redirectedPathname}?${this.setSortToQuery()}`;
    history.push(initialURL);
  }

  fetchData = async retry => {
    this.setURL(retry);
    try {
      const response = await fetch(`http://localhost:3000/characters?${this.setSortToQuery()}`);
      const { firstPage, lastPage } = await this.parseHeaders(response.headers);
      const characters = await response.json();
      await this.setState({
        characters,
        firstPage,
        lastPage
      });
      if (this.state.characters.length === 0 && this.state.currentPage !== '1') {
        this.setURL(true);
      }
    } catch (err) {
      console.log(err, err.stack);
    }
  }

  parseHeaders = headers => {
    const paginationHeaders = headers.get('Link') ? headers.get('Link').split(',') : []; // splitting header string into array
    const first = paginationHeaders[0];
    const last = paginationHeaders[2];
    const firstPage = first ? first.match(/page=(\d+)/)[1] : ''; // extracting first page number
    const lastPage = last ? last.match(/page=(\d+)/)[1] : ''; // extracting last page number
    return { firstPage, lastPage }
  }

  handleQueryChange = event => {
    const query = event.target ? event.target.value : '';
    const searchParam = query ? `&q=${query}` : '';
    this.setState({
      searchParam
    });
    this.fetchSearchData();
  }

  handleSortChange = async name => {
    const sortParam = name ? `&_sort=${name}&_order=asc` : '';
    await this.setState({
      sortParam
    });
    this.fetchData();
  }

  fetchSearchData = () => {
    this.fetchData();
  }

  deleteCharacter = async id => {
    const response = await fetch(`http://localhost:3000/characters/${id}`, {
      method: 'DELETE'
    });
    if (response.ok) {
      this.fetchSearchData();
    }
  }

  render() {
    const { params } = this.props.match;
    const { characters, currentPage, firstPage, lastPage } = this.state;
    const numCurrentPage = Number(params.page);
    const isFirstPage = currentPage === firstPage;
    const isLastPage = currentPage === lastPage;

    const paginationData = []
    for (let i = 1; i <= Number(lastPage); i++) {  // creating pagination items based on
      paginationData.push(                            // number of available pages from api
          <li key={'nav-li' + i} className="page-item">
          <NavLink
            to={{ pathname: `/characters/${i}`, search: `_page=${i}`}}
            className={`page-link ${numCurrentPage === i ? 'active' : ''}`}
          >
            {i} <span className="sr-only">{numCurrentPage === i ? '(current)' : ''}</span>
          </NavLink>
        </li>
      )
    }

    return (
      <main className="container">
        <h1>List View</h1>

        <div className="row">
          <div className="col-sm-6">
            <SearchInput
              id="searchInput"
              placeholder="Search characters..."
              label="Search characters..."
              handleQueryChange={this.handleQueryChange}
            />
          </div>
          <div className="col-sm-6 text-sm-right">
            <Link to='/add-character' className="btn btn-primary mb-3">Add New</Link>
          </div>
        </div>

        <ListItems 
          characters={characters}
          deleteCharacter={this.deleteCharacter}
          handleSortChange={this.handleSortChange}
        />

        {paginationData.length > 0 && <ListPagination
          isFirstPage={isFirstPage}
          isLastPage={isLastPage}
          numCurrentPage={numCurrentPage}
          paginationData={paginationData}
        />}
        {this.state.characters.length <= 0 && 
          <p>No Results Found</p>
        }
      </main>
    );
  }
}
export default ListView;