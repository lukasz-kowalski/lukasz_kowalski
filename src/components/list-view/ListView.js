import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import debounce from '../../utils/debounce';
import ListItems from './ListItems';
import ListPagination from './ListPagination';

import './ListView.css';

class ListView extends Component {
  constructor(props) {
    super(props);
    this.fetchQuery = debounce(this.fetchQuery.bind(this), 200);
  }
  state = {
    characters: [],
    currentPage: '',
    firstPage: '',
    lastPage: '',
    queryParam: ''
  }

  async componentDidMount() {
    this.fetchInitialData()
  }

  componentDidUpdate(prevProps) {
    if (prevProps.match.params.page !== this.props.match.params.page) {
      this.fetchInitialData()
    }
  }

  fetchInitialData = async () => {
    const { params } = this.props.match;  // get page from params
    const currentPage = params.page ? params.page : '1'; // if lack set to 1
    const pageParam = `_page=${currentPage}`;
    const response = await this.fetchData(pageParam);
    const { firstPage, lastPage } = await this.parseHeaders(response.headers)
    const characters = await response.json()
    this.setState({
      characters,
      currentPage,
      firstPage,
      lastPage
    })
  }

  fetchData = async params => {
    const response = await fetch(`http://localhost:3000/characters?${params}`);
    return response;
  }

  parseHeaders = headers => {
    const paginationHeaders = headers.get('Link').split(',') // splitting header string into array
    const first = paginationHeaders.filter(header => header.indexOf('first') > 0)[0]; // filtering array for first page
    const last = paginationHeaders.filter(header => header.indexOf('last') > 0)[0]; // filtering array for last page
    const firstPage = first ? first.slice(first.indexOf('=') + 1, first.indexOf('>')) : ''; // extracting first page number
    const lastPage = last ? last.slice(last.indexOf('=') + 1, last.indexOf('>')) : ''; // extracting last page number
    return { firstPage, lastPage }
  }

  handleChange = async event => {
    const query = event.target ? event.target.value : '';
    const queryParam = query ? `&q=${query}` : '';
    this.fetchQuery(queryParam);
  }

  fetchQuery = async queryParam => {
    if (queryParam) {
      const response = await this.fetchData(queryParam);
      const characters = await response.json();
      this.setState({
        characters,
        currentPage: '1',
        firstPage: '',
        lastPage: ''
      });
    } else {
      this.fetchInitialData();
    }
  }

  render() {
    const { characters ,currentPage, firstPage, lastPage } = this.state;
    const numCurrentPage = Number(currentPage)
    const isFirstPage = currentPage === firstPage;
    const isLastPage = currentPage === lastPage;

    const paginationData = []
    for (let i = 1; i <= Number(lastPage); i++) {  // creating pagination items based on
      paginationData.push(                            // number of available pages from api
          <li key={'nav-li' + i} className="page-item">
          <NavLink
            to={`/characters/${i}`}
            className={`page-link ${numCurrentPage === i && 'active'}`}
          >
            {i} <span className="sr-only">{numCurrentPage === i && '(current)'}</span>
          </NavLink>
        </li>
      )
    }

    return (
      <div className="container">
        <h1>List View</h1>

        <div className="row">
          <div className="col-sm-6">
            <div className="form-group">
              <label htmlFor="searchInput" className="sr-only">Search</label>
              <input 
                type="text" 
                className="form-control" 
                id="searchInput" 
                placeholder="Search..." 
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="col-sm-6 text-sm-right">
            <span className="btn btn-primary mb-3">Add New</span>
          </div>
        </div>

        <ListItems characters={characters} />            
        {paginationData.length > 0 && <ListPagination
          isFirstPage={isFirstPage}
          isLastPage={isLastPage}
          numCurrentPage={numCurrentPage}
          paginationData={paginationData} 
          queryParam={this.state.queryParam}
        />}
      </div>
    );
  }
}
export default ListView;