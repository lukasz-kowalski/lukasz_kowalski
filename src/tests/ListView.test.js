import React from 'react';
import { shallow } from 'enzyme';
import ListView from '../components/ListView/ListView';

let wrapper;

const match = {
  params: {
    page: '1'
  }
};

const location = {
  pathname: '/characters/1'
};

const history = {
  push: jest.fn()
};

const characters = [
  {
    "id": 1,
    "name": "Luke Skywalker",
    "species": "Human",
    "gender": "male",
    "homeworld": "Tatooine"
  },
  {
    "id": 2,
    "name": "C-3PO",
    "species": "Droid",
    "gender": "n/a",
    "homeworld": "Tatooine"
  }
];

const headers = {
  get: jest.fn(() => `<http://localhost:3000/characters?_page=1>; rel="first", 
  <http://localhost:3000/characters?_page=1>; rel="prev", 
  <http://localhost:3000/characters?_page=3>; rel="next", 
  <http://localhost:3000/characters?_page=3>; rel="last"`)
};

const headers2 = {
  get: jest.fn(() => `<http://localhost:3000/characters?_page=1&q=luke>; rel="first", 
  <http://localhost:3000/characters?_page=1&q=luke>; rel="prev", 
  <http://localhost:3000/characters?_page=5&q=luke>; rel="next", 
  <http://localhost:3000/characters?_page=5&q=luke>; rel="last"`)
};

beforeEach(() => {
  fetch.mockResponse(JSON.stringify(characters));
  wrapper = shallow(
  <ListView 
    match={match}
    location={location}
    history={history}
  />);
  wrapper.setState({
    searchParam: '&q=luke',
    sortParam: '&_sort=name&_order=asc'
  });
})

test('should successful render ListView with no results message', () => {
  expect(wrapper).toMatchSnapshot();
})

test('should successful render ListView with characters', () => {
  wrapper.setState({
    characters
  });
  expect(wrapper).toMatchSnapshot();
})

test('should return _page=1 string', () => {
  expect(wrapper.instance().getInitialQuery()).toBe('_page=1');
})

test('should return _page=1&q=luke string', () => {
  expect(wrapper.instance().setSearchToQuery()).toBe('_page=1&q=luke');
})

test('should return _page=1&q=luke&_sort=name&_order=asc string', () => {
  const query = '_page=1&q=luke&_sort=name&_order=asc'
  expect(wrapper.instance().setSortToQuery()).toBe(query);
})

test('should push history to /characters/1?_page=1 on first render', () => {
  wrapper.instance().setURL();
  expect(history.push.mock.calls[0][0]).toBe('/characters/1?_page=1');
})

test('should set characters state from api fetch', async () => {
  const mockedState = {
    characters,
    firstPage: '',
    lastPage: '',
    currentPage: '1',
    searchParam: '&q=luke',
    sortParam: '&_sort=name&_order=asc'
  };
  await wrapper.instance().fetchData();
  expect(wrapper.state()).toEqual(mockedState);
})

test('should parse header and return first and last page', () => {
  expect(wrapper.instance().parseHeaders(headers)).toEqual({ firstPage: '1', lastPage: '3' });
})

test('should parse header with query param and return first and last page', () => {
  expect(wrapper.instance().parseHeaders(headers2)).toEqual({ firstPage: '1', lastPage: '5' });
})

test('should change searchParam in component state to &q=luke', () => {
  const event = {
    target: {
      value: 'luke'
    }
  };
  wrapper.instance().handleQueryChange(event);
  expect(wrapper.state('searchParam')).toBe('&q=luke');
})

test('should change searchParam in component state to empty string when event has no value', () => {
  const event = {
    target: {
      value: ''
    }
  }
  wrapper.instance().handleQueryChange(event);
  expect(wrapper.state('searchParam')).toBe('');
})

test('should change sortParam in component state to &q=luke', () => {
  const name = 'species';
  const sortParam = '&_sort=species&_order=asc';
  wrapper.instance().handleSortChange(name);
  expect(wrapper.state('sortParam')).toBe(sortParam);
})

test('should change sortParam to empty string when no sort selected', () => {
  wrapper.instance().handleSortChange('');
  expect(wrapper.state('sortParam')).toBe('');
})
