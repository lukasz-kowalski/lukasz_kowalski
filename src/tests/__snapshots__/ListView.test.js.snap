// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should successful render ListView with characters 1`] = `
<main
  className="container"
>
  <h1>
    List View
  </h1>
  <div
    className="row"
  >
    <div
      className="col-sm-6"
    >
      <SearchInput
        handleQueryChange={[Function]}
        id="searchInput"
        label="Search characters..."
        placeholder="Search characters..."
      />
    </div>
    <div
      className="col-sm-6 text-sm-right"
    >
      <Link
        className="btn btn-primary mb-3"
        replace={false}
        to="/add-character"
      >
        Add New
      </Link>
    </div>
  </div>
  <ListItems
    characters={
      Array [
        Object {
          "gender": "male",
          "homeworld": "Tatooine",
          "id": 1,
          "name": "Luke Skywalker",
          "species": "Human",
        },
        Object {
          "gender": "n/a",
          "homeworld": "Tatooine",
          "id": 2,
          "name": "C-3PO",
          "species": "Droid",
        },
      ]
    }
    deleteCharacter={[Function]}
    handleSortChange={[Function]}
  />
</main>
`;

exports[`should successful render ListView with no results message 1`] = `
<main
  className="container"
>
  <h1>
    List View
  </h1>
  <div
    className="row"
  >
    <div
      className="col-sm-6"
    >
      <SearchInput
        handleQueryChange={[Function]}
        id="searchInput"
        label="Search characters..."
        placeholder="Search characters..."
      />
    </div>
    <div
      className="col-sm-6 text-sm-right"
    >
      <Link
        className="btn btn-primary mb-3"
        replace={false}
        to="/add-character"
      >
        Add New
      </Link>
    </div>
  </div>
  <ListItems
    characters={Array []}
    deleteCharacter={[Function]}
    handleSortChange={[Function]}
  />
  <p>
    No Results Found
  </p>
</main>
`;
